File: go/cluster_test.go
package flotilla

import (
	"strconv"
	"testing"
)

var defaultNodes = []string{"n1", "n2", "n3"}

func swarm(nGroups int, nodes []string) *Cluster {
	c := NewCluster(nodes)
	for g := 0; g < nGroups; g++ {
		c.AddGroup("g"+strconv.Itoa(g), nodes)
	}
	c.RunOnce()
	return c
}

func TestEveryGroupAgreesOnOneLeader(t *testing.T) {
	c := swarm(50, defaultNodes)
	leaders := c.Leaders()
	if len(leaders) != 50 {
		t.Fatalf("groups with a leader = %d, want 50", len(leaders))
	}
	for gid, who := range leaders {
		if len(who) != 1 {
			t.Fatalf("group %s disagreed on its leader: %v", gid, who)
		}
	}
}

func TestLeaderIsTheMinIDMember(t *testing.T) {
	c := swarm(10, defaultNodes)
	for gid, who := range c.Leaders() {
		if len(who) != 1 || !who["n1"] {
			t.Fatalf("group %s leader set = %v, want {n1}", gid, who)
		}
	}
}

func TestBatchingBeatsOneRPCPerGroup(t *testing.T) {
	// 200 groups across 3 nodes. Without batching the leader would emit
	// 200 * 2 = 400 separate RPCs; batching collapses each node's outbound to at
	// most (peers) envelopes, so envelopes << messages.
	c := swarm(200, defaultNodes)
	if c.TotalMessages() < 400 {
		t.Fatalf("totalMessages = %d, want >= 400", c.TotalMessages())
	}
	if c.TotalEnvelopes() > 6 {
		t.Fatalf("totalEnvelopes = %d, want <= 6", c.TotalEnvelopes())
	}
	if c.TotalEnvelopes() >= c.TotalMessages()/10 {
		t.Fatalf("envelopes %d not << messages %d", c.TotalEnvelopes(), c.TotalMessages())
	}
}

func TestScalesToThousandsOfGroups(t *testing.T) {
	c := swarm(2000, defaultNodes)
	leaders := c.Leaders()
	if len(leaders) != 2000 {
		t.Fatalf("groups with a leader = %d, want 2000", len(leaders))
	}
	for _, who := range leaders {
		if len(who) != 1 {
			t.Fatalf("a group disagreed on its leader: %v", who)
		}
	}
	if c.TotalEnvelopes() > 6 {
		t.Fatalf("totalEnvelopes = %d, want <= 6 even at 2000 groups", c.TotalEnvelopes())
	}
}

func TestEmptyClusterHasNoLeaders(t *testing.T) {
	c := swarm(0, defaultNodes)
	if len(c.Leaders()) != 0 {
		t.Fatalf("empty cluster should have no leaders, got %v", c.Leaders())
	}
	if c.TotalMessages() != 0 || c.TotalEnvelopes() != 0 {
		t.Fatalf("empty cluster sent traffic: messages=%d envelopes=%d", c.TotalMessages(), c.TotalEnvelopes())
	}
}

func TestSingleGroupSingleNodeSelfLeads(t *testing.T) {
	c := NewCluster([]string{"solo"})
	c.AddGroup("g0", []string{"solo"})
	c.RunOnce()
	leaders := c.Leaders()
	if len(leaders) != 1 || !leaders["g0"]["solo"] {
		t.Fatalf("solo node should lead its own group, got %v", leaders)
	}
	// A single-member group has no peers to notify, so nothing goes on the wire.
	if c.TotalMessages() != 0 || c.TotalEnvelopes() != 0 {
		t.Fatalf("solo group sent traffic: messages=%d envelopes=%d", c.TotalMessages(), c.TotalEnvelopes())
	}
}

func TestTwoNodesConvergeAfterDelivery(t *testing.T) {
	c := NewCluster([]string{"a", "b"})
	c.AddGroup("g0", []string{"a", "b"})
	c.RunOnce()
	// "a" is the min id, so both replicas should believe in "a".
	for _, sched := range c.Schedulers {
		if got := sched.Replicas["g0"].Leader; got != "a" {
			t.Fatalf("node %s believes leader is %q, want a", sched.NodeID, got)
		}
	}
	if c.TotalMessages() != 1 || c.TotalEnvelopes() != 1 {
		t.Fatalf("two-node group: messages=%d envelopes=%d, want 1 and 1", c.TotalMessages(), c.TotalEnvelopes())
	}
}

File: ts/src/cluster.test.ts
import { describe, expect, it } from "vitest";

import { Cluster } from "./cluster";

const NODES = ["n1", "n2", "n3"];

function swarm(nGroups: number, nodes: string[] = NODES): Cluster {
  const c = new Cluster(nodes);
  for (let g = 0; g < nGroups; g++) c.addGroup(`g${g}`, nodes);
  c.runOnce();
  return c;
}

describe("Cluster", () => {
  it("has every group agree on exactly one leader", () => {
    const c = swarm(50);
    const leaders = c.leaders();
    expect(leaders.size).toBe(50);
    for (const [gid, who] of leaders) {
      expect(who.size, `group ${gid} disagreed on its leader`).toBe(1);
    }
  });

  it("elects the min-id member as leader", () => {
    const c = swarm(10);
    for (const who of c.leaders().values()) {
      expect(who.size).toBe(1);
      expect(who.has("n1")).toBe(true); // min of n1,n2,n3
    }
  });

  it("beats one RPC per group through batching", () => {
    // 200 groups across 3 nodes. Without batching the leader would emit
    // 200 * 2 = 400 separate RPCs; batching collapses each node's outbound to at
    // most (peers) envelopes, so envelopes << messages.
    const c = swarm(200);
    expect(c.totalMessages()).toBeGreaterThanOrEqual(400);
    expect(c.totalEnvelopes()).toBeLessThanOrEqual(6); // <= 3 nodes * 2 peers
    expect(c.totalEnvelopes()).toBeLessThan(c.totalMessages() / 10);
  });

  it("scales to thousands of groups with bounded envelopes", () => {
    const c = swarm(2000);
    const leaders = c.leaders();
    expect(leaders.size).toBe(2000);
    expect([...leaders.values()].every((w) => w.size === 1)).toBe(true);
    // Envelope count stays bounded by peer pairs even at 2000 groups.
    expect(c.totalEnvelopes()).toBeLessThanOrEqual(6);
  });

  it("has no leaders and no traffic when empty", () => {
    const c = swarm(0);
    expect(c.leaders().size).toBe(0);
    expect(c.totalMessages()).toBe(0);
    expect(c.totalEnvelopes()).toBe(0);
  });

  it("lets a single node lead its own group without traffic", () => {
    const c = new Cluster(["solo"]);
    c.addGroup("g0", ["solo"]);
    c.runOnce();
    const leaders = c.leaders();
    expect(leaders.size).toBe(1);
    expect(leaders.get("g0")?.has("solo")).toBe(true);
    expect(c.totalMessages()).toBe(0);
    expect(c.totalEnvelopes()).toBe(0);
  });

  it("converges two nodes after a single delivery", () => {
    const c = new Cluster(["a", "b"]);
    c.addGroup("g0", ["a", "b"]);
    c.runOnce();
    for (const sched of c.schedulers.values()) {
      expect(sched.replicas.get("g0")?.leader).toBe("a");
    }
    expect(c.totalMessages()).toBe(1);
    expect(c.totalEnvelopes()).toBe(1);
  });
});
